### PR TITLE
[OPERATOR-507] - implement daemonset backup

### PR DESCRIPTION
**Pull request by haibinxie**
**What this PR does / why we need it**:
backup portworx components that are deployed by daemonset method, if migration to operator fails we can rollback.
https://portworx.atlassian.net/browse/OPERATOR-507

**Which issue(s) this PR fixes** (optional)
Closes #

**Special notes for your reviewer**:

### CODE REVIEW

**Review by piyush-nimbalkar — COMMENTED** [review]
Overall the change looks good.

File: pkg/migration/backup.go [hunk 1]
```diff
@@ -0,0 +1,284 @@
...
+
+	var sb strings.Builder
+	for _, obj := range objs {
+		bytes, err := yaml.Marshal(obj)
+		if err != nil {
+			return err
+		}
+
+		sb.WriteString(string(bytes))
+		sb.WriteString("\n")
```
[the remark this answers is not in the document]
Comment (haibinxie) [reply pkg/migration/backup.go:64]: I dont think so. it actually worked without it.

File: pkg/migration/migration.go [hunk 2]
```diff
@@ -212,6 +218,8 @@ func (h *Handler) processMigration(
 
 	// TODO: Wait for all components to be up, before marking the migration as completed
 
+	// TODO: once daemonset is deleted, if we restart operator all code after this line
```
[the remark this answers is not in the document]
Comment (haibinxie) [reply pkg/migration/migration.go:221]: right now there is unmarkAllDoneNodes and TODOs.

File: pkg/migration/backup_test.go [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+
+	go migrator.Start()
+	time.Sleep(10 * time.Second)
+
+	// Verify the configmap is created
+	cm := v1.ConfigMap{}
+	err := testutil.Get(k8sClient, &cm, BackupConfigMapName, ds.Namespace)
+	require.NoError(t, err)
+
+	fmt.Printf("%v\n", cm.Data)
```
[the remark this answers is not in the document]
Comment (haibinxie) [reply pkg/migration/backup_test.go:107]: I should try automatic rollback after this line. the test ends here in this PR for manual verification of rollback.

File: pkg/migration/migration.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (piyush-nimbalkar) [reply pkg/migration/migration.go:221]: Actually we cannot unmark the nodes before deleting the daemonset, else the daemonset will start creating it's pods again. Just something to remember for future.

File: pkg/migration/backup_test.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (haibinxie) [reply pkg/migration/backup_test.go:107]: we can implement rollback after FC, it requires non-trivial efforts

File: pkg/migration/backup.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (haibinxie) [reply pkg/migration/backup.go:64]: I was surprised too, after some more test, looks like kubectl does not throw an error, but without splitter it does not apply all objects

File: pkg/migration/backup.go [hunk 4]
```diff
@@ -0,0 +1,298 @@
...
+
+		sb.WriteString(string(bytes))
+		sb.WriteString("---\n")
+	}
+
+	data := map[string]string{
+		"backup": sb.String(),
+	}
+
+	return h.client.Create(context.TODO(),
```
Comment (piyush-nimbalkar) [inline pkg/migration/backup.go:85]: Let's not fail on IsAlreadyPresent error, else on restart the migration will get stuck here. 
Let's print that the backup has already been taken and return nil.

File: pkg/migration/backup_test.go [hunk 5]
```diff
@@ -0,0 +1,154 @@
...
+)
+
+func TestBackup(t *testing.T) {
+	testBackup(t, true)
+	testBackup(t, false)
+}
+
+func testBackup(t *testing.T, backupExits bool) {
+	clusterName := "px-cluster"
+	ds := &appsv1.DaemonSet{
```
Comment (piyush-nimbalkar) [inline pkg/migration/backup_test.go:33]: Ideally we would want to check for all objects that are backed up, but we can add them later when we actually verify those objects from the config map.

**Review by piyush-nimbalkar — APPROVED** [review]
LGTM after just one comment.

File: pkg/migration/backup.go (hunk 4 above)
Comment (haibinxie) [reply pkg/migration/backup.go:85]: this is already checked at the beginning of the function, we already return if configmap exists.

Comment (piyush-nimbalkar) [reply pkg/migration/backup.go:85]: Oh sorry, missed that.

**Pull request merged**